File: app/src/main/java/si/roglan/EMP_Seminarska/Nakup.java
package si.roglan.EMP_Seminarska;


import android.content.Context;
import android.os.Bundle;
import android.support.design.widget.Snackbar;
import android.support.v4.app.Fragment;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AutoCompleteTextView;
import android.widget.Button;
import android.widget.CheckBox;
import android.widget.CompoundButton;
import android.widget.EditText;
import android.widget.RelativeLayout;
import android.widget.Spinner;
import android.widget.TextView;

import com.google.android.gms.common.api.GoogleApiClient;
import com.google.android.gms.common.api.Status;
import com.google.android.gms.location.places.AutocompleteFilter;
import com.google.android.gms.location.places.GeoDataClient;
import com.google.android.gms.location.places.Place;






//import com.google.android.gms.location.places.ui.SupportPlaceAutocompleteFragment ;


import com.google.android.gms.location.places.PlaceDetectionClient;
import com.google.android.gms.location.places.Places;
import com.google.android.gms.location.places.ui.PlaceAutocompleteFragment;
import com.google.android.gms.location.places.ui.PlaceSelectionListener;

import org.w3c.dom.Text;

import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Locale;


public class Nakup extends Fragment {
    protected GeoDataClient mGeoDataClient;
    private GoogleApiClient mGoogleApiClient;

    HomeListener activityCommander;
    private RelativeLayout home_container;
    private Spinner destinacije;
    private EditText datum_odhoda;
    private Spinner st_oseb_spinner;
    private Spinner razred_odhoda;
    private CheckBox dvosmerna;
    private EditText datum_prihoda;
    private Spinner razred_prihoda;
    private TextView datum_prihoda_view;
    private TextView razred_prihoda_view;
    private TextView paste_location_view;

    public Nakup() {
        // Required empty public constructor
    }

    @Override
    public void onAttach(Context context) {
        try {
            activityCommander = (HomeListener) context;
        } catch (ClassCastException e) {
            throw new ClassCastException(context.toString());
        }
        super.onAttach(context);
    }


    public void recieveUserData(Bundle bundle) {
        if (bundle != null) {

            ArrayList<String> userData = bundle.getStringArrayList("userData");
            if (userData != null) {
                /*destinacije.setText(userData[0]);
                home_surname_input_var.setText(userData[1]);
                if(userData[2].equals("Moški")) radioButton_male_var.setChecked(true);
                else radioButton_female_var.setChecked(true);
                visina_input_var.setText(userData[3]);
                */
                Snackbar.make(home_container, userData.get(0) + "<- ---USER DATA 0", Snackbar.LENGTH_LONG).show();
            }
        }
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_nakup, container, false);


        home_container = (RelativeLayout) view.findViewById(R.id.home_container);

        destinacije = (Spinner) view.findViewById(R.id.destinacije_spinner);
        datum_odhoda = (EditText) view.findViewById(R.id.datum_odhoda_field);
        st_oseb_spinner = (Spinner) view.findViewById(R.id.st_oseb_spinner);


        razred_odhoda = (Spinner) view.findViewById(R.id.razred_odhoda_spinner);
        dvosmerna = (CheckBox) view.findViewById(R.id.checkBox);
        datum_prihoda = (EditText) view.findViewById(R.id.datum_prihoda);
        razred_prihoda = (Spinner) view.findViewById(R.id.razredPrihoda);


        datum_prihoda_view = (TextView) view.findViewById(R.id.datum_prihoda_view);
        razred_prihoda_view = (TextView) view.findViewById(R.id.razred_prihoda_view);


        datum_prihoda.setVisibility(View.GONE);
        razred_prihoda.setVisibility(View.GONE);
        datum_prihoda_view.setVisibility(View.GONE);
        razred_prihoda_view.setVisibility(View.GONE);


        recieveUserData(this.getArguments());


        final Button doloci_potnike_button = (Button) view.findViewById(R.id.doloci_potnike_button);
        doloci_potnike_button.setOnClickListener( //listener
                new View.OnClickListener() {
                    public void onClick(View v) {
                        dolociPotnikeButtonClicked(v);
                    }
                }
        );

        dvosmerna.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                if (isChecked) {
                    datum_prihoda.setVisibility(View.VISIBLE);
                    razred_prihoda.setVisibility(View.VISIBLE);
                    datum_prihoda_view.setVisibility(View.VISIBLE);
                    razred_prihoda_view.setVisibility(View.VISIBLE);
                } else {

                    datum_prihoda.setVisibility(View.GONE);
                    razred_prihoda.setVisibility(View.GONE);
                    datum_prihoda_view.setVisibility(View.GONE);
                    razred_prihoda_view.setVisibility(View.GONE);
                }
                // do something, the isChecked will be
                // true if the switch is in the On position
            }
        });






        /*SupportPlaceAutocompleteFragment  autocompleteFragment = (SupportPlaceAutocompleteFragment )
                getFragmentManager().findFragmentById(R.id.place_autocomplete_fragment);

                */

        // Construct a GeoDataClient.
        mGeoDataClient = Places.getGeoDataClient(getActivity(), null);

        // Construct a PlaceDetectionClient.
        PlaceDetectionClient mPlaceDetectionClient = Places.getPlaceDetectionClient(getActivity(), null);

        mGoogleApiClient = new GoogleApiClient
                .Builder(getActivity())
                .addApi(Places.GEO_DATA_API)
                .addApi(Places.PLACE_DETECTION_API)
                .enableAutoManage(getActivity(), null)
                .build();


        PlaceAutocompleteFragment autocompleteFragment = (PlaceAutocompleteFragment)
                getActivity().getFragmentManager().findFragmentById(R.id.place_autocomplete_fragment);



        paste_location_view =(TextView) view.findViewById(R.id.paste_location_view);

        if(autocompleteFragment != null) {

            autocompleteFragment.setOnPlaceSelectedListener(new PlaceSelectionListener() {
                public static final String TAG = "bla";

                @Override
                public void onPlaceSelected(Place place) {
                    // TODO: Get info about the selected place.
                    Log.i(TAG, "Place: " + place.getName());

                    String placeDetailsStr = place.getName() + "\n"
                            + place.getId() + "\n"
                            + place.getLatLng().toString() + "\n"
                            + place.getAddress() + "\n"
                            + place.getAttributions();
                    paste_location_view.setText(placeDetailsStr);
                }

                @Override
                public void onError(Status status) {
                    // TODO: Handle the error.
                    Log.i(TAG, "An error occurred: " + status);
                }
            });
        }else{
            Log.i("blyat", " NULLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLL " );
        }




        return view;
    }

    private boolean checkFields() {

        if (datum_odhoda.getText().toString().equals("")) {
            Snackbar.make(home_container, "Vnesite datum odhoda", Snackbar.LENGTH_LONG).show();
            return false;
        }
        try {
            DateFormat format = new SimpleDateFormat("d.m.yyyy", Locale.ENGLISH);
            format.parse(datum_odhoda.getText().toString());

        } catch (Exception e) {
            Snackbar.make(home_container, "Vnesite pravilen format datum odhoda -> dan.mesec.leto", Snackbar.LENGTH_LONG).show();
            return false;
        }


        if (dvosmerna.isChecked()) {
            if (datum_prihoda.getText().toString().equals("")) {
                Snackbar.make(home_container, "Vnesite datum prihoda", Snackbar.LENGTH_LONG).show();
                return false;
            }
            try {
                DateFormat format = new SimpleDateFormat("d.m.yyyy", Locale.ENGLISH);
                format.parse(datum_prihoda.getText().toString());
            } catch (Exception e) {
                Snackbar.make(home_container, "Vnesite pravilen format datum prihoda -> dan.mesec.leto", Snackbar.LENGTH_LONG).show();
                return false;
            }
        }


        return true;
    }
















    private void dolociPotnikeButtonClicked(View view) { //gets called when button is clicked


        if (!checkFields())
            return;

        Button b = (Button) view;
        String button_label = b.getText().toString();

        ArrayList<String> nakup_data = new ArrayList<>();
        nakup_data.add(destinacije.getSelectedItem().toString());
        nakup_data.add(datum_odhoda.getText().toString());
        nakup_data.add(st_oseb_spinner.getSelectedItem().toString());
        nakup_data.add(razred_odhoda.getSelectedItem().toString());

        if (dvosmerna.isChecked()) {
            nakup_data.add(datum_prihoda.getText().toString());
            nakup_data.add(razred_prihoda.getSelectedItem().toString());
        }
        activityCommander.sendNakupData(nakup_data);


        if (button_label.equals("DOLOČI POTNIKE")) {
            activityCommander.setPotnikiFragment();
        } else {
            activityCommander.setPlaciloFragment(); //passamo input
        }


    }


    interface HomeListener {
        void sendNakupData(ArrayList<String> userData);

        void setPlaciloFragment();

        void setPotnikiFragment();
    }


}

File: app/src/main/java/si/roglan/EMP_Seminarska/Nakup.kt
package si.roglan.EMP_Seminarska


import android.app.DatePickerDialog
import android.content.Context
import android.os.Bundle
import android.support.design.widget.Snackbar
import android.support.v4.app.Fragment
import android.support.v4.app.FragmentTransaction
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*
import com.google.android.gms.common.api.GoogleApiClient
import com.google.android.gms.common.api.Status
import com.google.android.gms.location.places.GeoDataClient
import com.google.android.gms.location.places.Place
import com.google.android.gms.location.places.Places
import com.google.android.gms.location.places.ui.PlaceAutocompleteFragment
import com.google.android.gms.location.places.ui.PlaceSelectionListener
import kotlinx.android.synthetic.main.fragment_nakup.*
import java.text.SimpleDateFormat
import java.util.*
import com.google.android.gms.location.places.ui.SupportPlaceAutocompleteFragment
import android.widget.EditText
import android.widget.RadioButton
import android.widget.DatePicker






class Nakup : Fragment() {
    private lateinit var mGeoDataClient: GeoDataClient
    private var mGoogleApiClient: GoogleApiClient? = null
    private lateinit var activityCommander: NakupListener

    private lateinit var datum_prihoda: EditText
    private lateinit var razred_prihoda: Spinner
    private lateinit var datum_prihoda_view: TextView
    private lateinit var razred_prihoda_view: TextView
    private lateinit var dvosmerna: CheckBox

    private var mOdhodAutocompleteString: String = ""
    private var mPrihodAutocompleteString: String = ""
    private var mOdhodAutocomplete = SupportPlaceAutocompleteFragment()
    private var mPrihodAutocomplete = SupportPlaceAutocompleteFragment()


    override fun onAttach(context: Context?) {
        try {
            activityCommander = (context as NakupListener?)!!
        } catch (e: ClassCastException) {
            throw ClassCastException(context.toString())
        }

        super.onAttach(context)
    }


    fun recieveNakupData(bundle: Bundle?) {
        if (bundle != null) {

            val nakupData = bundle.getStringArrayList("nakupData")
            if (nakupData != null) {

                /*nakupData.add(fromLocation)
                nakupData.add(toLocation)
                nakupData.add(date)
                nakupData.add(travelClass)
                nakupData.add(returnDate)
                nakupData.add(returnClass)*/

                mOdhodAutocompleteString = nakupData[0]
                mPrihodAutocompleteString = nakupData[1]

            }
        }
    }

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val view = inflater!!.inflate(R.layout.fragment_nakup, container, false)
        datum_prihoda = view.findViewById(R.id.datum_prihoda) as EditText
        razred_prihoda = view.findViewById(R.id.razred_prihoda) as Spinner
        datum_prihoda_view = view.findViewById(R.id.datum_prihoda_view) as TextView
        razred_prihoda_view = view.findViewById(R.id.razred_prihoda_view) as TextView
        dvosmerna = view.findViewById(R.id.dvosmerna) as CheckBox


        datum_prihoda.visibility = View.GONE
        razred_prihoda.visibility = View.GONE
        datum_prihoda_view.visibility = View.GONE
        razred_prihoda_view.visibility = View.GONE


        //Set initial departure and arrival dates
        val c = Calendar.getInstance();
        val df = SimpleDateFormat("dd.MM.yyyy");

        var datum_odhoda = view.findViewById(R.id.datum_odhoda) as EditText
        datum_odhoda.setText(df.format(c.getTime()));

        c.add(Calendar.DATE, 1);
        datum_prihoda.setText(df.format(c.getTime()));


        dvosmerna.setOnCheckedChangeListener { buttonView, isChecked ->
            if (isChecked) {
                datum_prihoda.visibility = View.VISIBLE
                razred_prihoda.visibility = View.VISIBLE
                datum_prihoda_view.visibility = View.VISIBLE
                razred_prihoda_view.visibility = View.VISIBLE

                //Set valid return date
                var datumOdhoda = view.findViewById(R.id.datum_odhoda) as EditText;
                val dateArray = datumOdhoda!!.text.toString().split("\\.".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()

                val c = Calendar.getInstance()
                c.set(Calendar.DAY_OF_MONTH, Integer.parseInt(dateArray[0]))
                c.set(Calendar.MONTH, Integer.parseInt(dateArray[1]) - 1)
                c.set(Calendar.YEAR, Integer.parseInt(dateArray[2]))
                c.add(Calendar.DATE, 1)
                datum_prihoda!!.setText(SimpleDateFormat("dd.MM.yyyy").format(c.time))

            } else {
                datum_prihoda.visibility = View.GONE
                razred_prihoda.visibility = View.GONE
                datum_prihoda_view.visibility = View.GONE
                razred_prihoda_view.visibility = View.GONE
            }
        }


        datum_odhoda!!.setOnClickListener(object: View.OnClickListener {
            override fun onClick(v : View?){
                val dateArray = datum_odhoda.text.toString().split("\\.".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
                val day = Integer.parseInt(dateArray[0])
                val month = Integer.parseInt(dateArray[1])
                val year = Integer.parseInt(dateArray[2])

                val dpd = DatePickerDialog(context,
                        DatePickerDialog.OnDateSetListener { view, year, monthOfYear, dayOfMonth ->
                            datum_odhoda!!.setText(
                                    String.format("%02d", dayOfMonth) + "." +
                                            String.format("%02d", monthOfYear + 1) + "." +
                                            String.format("%02d", year)
                            )

                            if (dvosmerna.isChecked) {
                                val dateArray = datum_prihoda!!.text.toString().split("\\.".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
                                val returnDay = Integer.parseInt(dateArray[0])
                                val returnMonth = Integer.parseInt(dateArray[1])
                                val returnYear = Integer.parseInt(dateArray[2])

                                if (year > returnYear || monthOfYear + 1 > returnMonth || dayOfMonth >= returnDay) {
                                    val c = Calendar.getInstance()
                                    c.set(Calendar.DAY_OF_MONTH, dayOfMonth)
                                    c.set(Calendar.MONTH, monthOfYear)
                                    c.set(Calendar.YEAR, year)
                                    c.add(Calendar.DATE, 1)
                                    datum_prihoda!!.setText(SimpleDateFormat("dd.MM.yyyy").format(c.time))
                                }
                            }
                        }, year, month - 1, day)

                //Limit date to today
                val c = Calendar.getInstance()
                dpd.datePicker.minDate = c.time.time
                dpd.show()
            }
        })


        datum_prihoda!!.setOnClickListener(object: View.OnClickListener {
            override fun onClick(v: View?) {
                val dateArray = datum_prihoda.text.toString().split("\\.".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
                val day = Integer.parseInt(dateArray[0])
                val month = Integer.parseInt(dateArray[1])
                val year = Integer.parseInt(dateArray[2])

                val dpd = DatePickerDialog(context,
                    DatePickerDialog.OnDateSetListener { view, year, monthOfYear, dayOfMonth ->
                        val fday = String.format("%02d", dayOfMonth)
                        val fmonth = String.format("%02d", monthOfYear + 1)
                        val fyear = String.format("%02d", year)
                        datum_prihoda.setText("$fday.$fmonth.$fyear")
                    }, year, month - 1, day)


                val ddateArray = datum_prihoda.text.toString().split("\\.".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
                val limitDay = Integer.parseInt(ddateArray[0]);
                val limitMonth = Integer.parseInt(ddateArray[1]);
                val limitYear = Integer.parseInt(ddateArray[2]);

                val c = Calendar.getInstance();
                c.set(Calendar.DAY_OF_MONTH, limitDay);
                c.set(Calendar.MONTH, limitMonth - 1);
                c.set(Calendar.YEAR, limitYear);
                dpd.getDatePicker().setMinDate(c.getTime().getTime());
                dpd.show();
            }
        })

        setupAutocompleteFragments(view)

        //Must be called after setting autocomplete fragments (avoid null values)
        val doloci_potnike_button = view.findViewById(R.id.doloci_potnike_button) as Button
        doloci_potnike_button.setOnClickListener { v -> dolociPotnikeButtonClicked(v) }

        recieveNakupData(this.arguments)

        return view
    }



    override fun onStop() {
        super.onStop()

        mGoogleApiClient!!.stopAutoManage(activity)
        mGoogleApiClient!!.disconnect()
    }

    private fun setAutocompleteFragmentOnSelectedListener(fragment: SupportPlaceAutocompleteFragment){

    }

    private fun setupAutocompleteFragments(view: View) {


        // Construct a GeoDataClient.
        mGeoDataClient = Places.getGeoDataClient(activity, null)

        // Construct a PlaceDetectionClient.
        val mPlaceDetectionClient = Places.getPlaceDetectionClient(activity, null)

        if (mGoogleApiClient == null)
        mGoogleApiClient = GoogleApiClient.Builder(activity)
                .addApi(Places.GEO_DATA_API)
                .addApi(Places.PLACE_DETECTION_API)
                .enableAutoManage(activity, null)
                .build()
        else
            mGoogleApiClient!!.connect()


        mOdhodAutocomplete.setOnPlaceSelectedListener(object : PlaceSelectionListener {
            override fun onPlaceSelected(place: Place) {
                Log.i("Autocomplete", "Place: " + place.name)

                /*mOdhodAutocompleteString = (place.name.toString() + "\n"
                        + place.id + "\n"
                        + place.latLng.toString() + "\n"
                        + place.address + "\n"
                        + place.attributions)*/
                //paste_location_view.text = placeDetailsStr

                mOdhodAutocompleteString = place.name.toString()
                mOdhodAutocomplete.setText(mOdhodAutocompleteString)
            }

            override fun onError(status: Status) {
                Log.i("Autocomplete fragment", "An error occurred: " + status)
            }
        })

        mPrihodAutocomplete.setOnPlaceSelectedListener(object : PlaceSelectionListener {
            override fun onPlaceSelected(place: Place) {
                Log.i("Autocomplete", "Place: " + place.name)

                mPrihodAutocompleteString = place.name.toString()
                mPrihodAutocomplete.setText(mPrihodAutocompleteString)
            }

            override fun onError(status: Status) {
                Log.i("Autocomplete fragment", "An error occurred: " + status)
            }
        })


        val ft = fragmentManager.beginTransaction()
        ft.replace(R.id.fragment_odhod, mOdhodAutocomplete)
        ft.replace(R.id.fragment_prihod, mPrihodAutocomplete)
        ft.commit()
    }

    private fun checkFields(): Boolean {

        if (datum_odhoda.text.toString() == "") {
            Snackbar.make(home_container, "Vnesite datum odhoda", Snackbar.LENGTH_LONG).show()
            return false
        }
        try {
            val format = SimpleDateFormat("d.m.yyyy", Locale.ENGLISH)
            format.parse(datum_odhoda.text.toString())

        } catch (e: Exception) {
            Snackbar.make(home_container, "Vnesite pravilen format datum odhoda -> dan.mesec.leto", Snackbar.LENGTH_LONG).show()
            return false
        }


        if (dvosmerna.isChecked) {
            if (datum_prihoda.text.toString() == "") {
                Snackbar.make(home_container, "Vnesite datum prihoda", Snackbar.LENGTH_LONG).show()
                return false
            }
            try {
                val format = SimpleDateFormat("d.m.yyyy", Locale.ENGLISH)
                format.parse(datum_prihoda.text.toString())
            } catch (e: Exception) {
                Snackbar.make(home_container, "Vnesite pravilen format datum prihoda -> dan.mesec.leto", Snackbar.LENGTH_LONG).show()
                return false
            }

        }


        return true
    }


    private fun dolociPotnikeButtonClicked(view: View) { //gets called when button is clicked


        //if (!checkFields()) // TODO COMMENTED CHECKING ON FRAGMENT NAKUP - TO DEBUG EASIER
        //    return;


        val button_label = (view as Button).text.toString()


        val nakup_data = ArrayList<String>()
        nakup_data.add(mOdhodAutocompleteString)
        nakup_data.add(mPrihodAutocompleteString)
        nakup_data.add(datum_odhoda.text.toString())
        nakup_data.add(razred_odhoda.selectedItem.toString())

        if (dvosmerna.isChecked) {
            nakup_data.add(datum_prihoda.text.toString())
            nakup_data.add(razred_prihoda.selectedItem.toString())
        }
        activityCommander.sendNakupData(nakup_data)


        if (button_label == "DOLOČI POTNIKE") {
            activityCommander.setPotnikiFragment()
        } else {
            activityCommander.setPlaciloFragment() //passamo input
        }


    }


    internal interface NakupListener {
        fun sendNakupData(userData: ArrayList<String>)

        fun setPlaciloFragment()

        fun setPotnikiFragment()
    }
}